### ISSUE TITLE
Non-static version of Mapper

### ISSUE THREAD

**Issue by jbogard** [AutoMapper/AutoMapper#1031]
Things to do:
- [x] Introduce non-static API
- [x] Obsolete existing static API
- [x] Provide bridge for LINQ #1038 
- [x] Read-only configuration #1039

I want to move to a non-static version of AutoMapper, that would get rid of so many headaches on my part of threading and junk. I'm thinking of something like:

```
var mapper = new Mapper(cfg => cfg.CreateMap<Source, Dest>());

var dest = mapper.Map<Source, Dest>(new Source());
```

I'd have overloads for the ctor:

```
public Mapper(Action<IConfiguration> configure);

public Mapper(IConfigurationProvider configuration);
```

And the configuration provider, if you wanted to initialize that by itself, could have a ctor too:

```
public class Configuration(Action<IConfigurationProvider> configure) {
   configure(this);
}
```

Technically configuration is still mutable after Seal(), since you can still go into TypeMaps and do whatever, grab the instance of the configuration and so on.

# Motivation

The static has been haunting me for years. Threading issues, order-of-operations, I can't fix that until I get rid of the statics. The changes in inheritance in 4.x have opened enough issues that I want to be deliberate about this move.

# Upgrade path

If you're using `Mapper.Initialize`, it's a matter of replacing the outer path with the real thing. I want to use the `Mapper` class and a `Configuration` class. I would obsolete all the static methods, obsolete the MappingEngine class and make the Mapper class inherit from it for 4.2. Then in 5.0, remove all the static.

Users, instead of calling Mapper.Map, would use a mapper instance directly, which they can make static all they like. Or configure in a container, whatever.

# Static compromise

~~Another option is to keep `Mapper.Initialize` and `Mapper.Map` and nothing else. Still include static field for the mapper and the configuration instances. I would still remove the static keyword from Mapper and move the MappingEngine pieces inside of Mapper, with the clear signal that "you might want to use the instance methods".~~ 

I want to move to a class called "Mapper" and in C# a static method hides an instance method of the same name. It's basically static or nothing. I could do something like Mapper.Engine.Map but that seems hokey.

@TylerCarlson1 @lbargaoanu thoughts?

**Comment by flq** [comment]
Would the Mapper instance still be usable by multiple threads?

**Comment by jbogard** [comment]
@flq Yes, I would continue to use concurrent dictionaries to cache mapping plans.

**Comment by TylerCarlson1** [comment]
In the long run I would say it would be best to go into instances of Mapper and non static option.

For IConfigurationProvider hide seal, they shouldn't have the option to call it themselves, it should call when done configuration action.

Give people options to do their unconventional un-supported stuff better.
- Conditional mappings can have multiple mappers for each condition. (Can resolve from IoC or pass a list of mappers)
- Can Mock and berate Automapper for unit tests (IMapper interface needed though)

As far as IMapper goes could I suggest just having 
`object Map(object source, Type sourceType, Type destinationType)` & 
`object Map(object source, Type sourceType, Type destinationType,
            Action<IMappingOperationOptions> opts)`
and have extensions from IMapper that do the generic implementations with documentation.
That way the mockers only have to mock 2 function calls instead of all the functions which call these two functions.
This is a personal preference of mine, with interfaces and overriding function calls.  There might be reasons I'm not aware of that would cause problems by doing this, other than accidentally mocking extension methods.  I can make a PR showing example on how I would implement this.
Also IMapper shouldn't have anything but Map functions I think, cause that's all it should be doing.  The assert configuration is valid and other things should be left in the configuration and/or configurationprovider.

There might be other things I would like to remove, or put in debug view only.  Like `FindTypeMap` or `GetAllTypeMaps`, because like you said, you have the option to edit them afterwards, but I have to think about it more.  Some of these things might not have as much use anymore, with the addition of conventions.

**Comment by hazzik** [comment]
I would like to propose following syntax:

```
var conf = new MapperConfiguration();
conf.AddMap<Source, Destination>();
var mapper = conf.CreateMapper(); // or BuildMapper();
```

**Comment by TylerCarlson1** [comment]
@hazzik I think that will have the same problem as a static instance has now, unless you want `conf.BuildMapper()` to clone all the mappings from mapper configuration and seal them.

If you don't clone them and add maps and call BuildMapper again you might get different functionality than if you added it before BuildMapper, which means order matters and is one of the problems with a static instance.

If AddMap is not based on CreateMap and is a wrapper to generate CreateMap when building the mapper, then you have to make a whole new set of classes and functions to lazily generate the mappings, which is a lot more code changes.

I like the idea of how you can make slightly different mappers easier this way.  Would have to duplicate configuration code otherwise.  But I feel it's going to be a lot of hard work trying to stop missuses and new issues that Initialize is causing right now.

**Comment by jbogard** [comment]
I'm not sure it will have problems but @hazzik I would likely keep the same config syntax (though I can certainly change type names). ConfigurationStore was always a weird name anyway, I'm also looking at "fixing" the names here (since Mapper would be the new MappingEngine).

**Comment by lbargaoanu** [comment]
About generic methods. People hate when they explicitly state generic type parameters and then we ignore that and use GetType :)
About threading. Having an instance would allow thread safe publication. So you configure the instance on one thread and then you have an explicit step of publishing for every thread. You can write it to a static or Register it with a container. The simplest way to do that is to always construct a "finished", read-only instance. So only Map in the interface. And then it should be only about caching, which has to be thread safe. Threading doesn't have to be messy when you're deliberate about it and you stick to known patterns.

**Comment by jbogard** [comment]
I don't think I'd remove the overloads, I'm trying to minimize the number of changes. So you'd just replace static calls with instance ones.

**Comment by jbogard** [comment]
I started a working example in #1034

**Comment by jbogard** [comment]
This is what I have for configuration:

```
var config = new MapperConfiguration(cfg => {
    cfg.CreateMap<Source, Dest>();
});
var config = new MapperConfiguration();
config.CreateMap<Source, Dest>();
```

Once you have your configuration, you use the configuration object to build a mapper:

```
var mapper = config.CreateMapper();
mapper.Map<Source, Dest>(foo);
```

Or for LINQ:

```
var builder = config.CreateExpressionBuilder();
var dtos = dbContext.Users.ProjectTo<UserDto>(builder).ToList();
```

It's a little weird that the config object has CreateMap and CreateMapper. Also, I'm now forcing a parameter to the LINQ methods. I don't think that's a huge deal, but I could expose a static setter somewhere:

```
var builder = config.CreateExpressionBuilder();
AutoMapper.QueryableExtensions.Factories.ExpressionBuilderFactory = () => builder;
var dtos = dbContext.Users.ProjectTo<UserDto>().ToList();
```

The two open items are the CreateMapper naming and whether or not to include some sort of factory callback for simplifying the LINQ piece. I'm not too keen on that because I don't want to proliferate static state.

**Comment by TylerCarlson1** [comment]
My concern with
```
var config = new MapperConfiguration();
config.CreateMap<Source, Dest>();
```

Is that you have call `config.Seal();` in order to guarantee you don't have order of operations problems like you did with the static API.  Well at this point you should be able to take every Issue that was solved by using `Mapper.Seal()` or `Mapper.Initalize()` and be able to re-create the exact same issue.

**Comment by jbogard** [comment]
So in that case, config.CreateMapper() calls Seal()

**Comment by TylerCarlson1** [comment]
If I create map of class that inherits, then class that derives from parent, with include.  Wouldn't the first map be sealed, so it wouldn't add the included derived types of the derived class mappings?
Or you map the other way around with mapping the parent with As<> or IncludeBase before mapping the base class, and the seal tries to find the base map but can't.  Would it possibly throw an error or silently fail and include functionality doesn't work?

**Comment by jbogard** [comment]
Ah, it could. In that case I'd only allow the configuration to be set via the constructor via that lambda thingy, and the actual MapperConfiguration object to only expose the IConfigurationProvider stuff.

If people are doing Mapper.CreateMap, that'll be harder for them to convert, since they'll have to gather them all up. But I suppose that's best to force you to define all the config up front.

**Comment by TylerCarlson1** [comment]
Yea I thought that was one of the main points to going to instances.  It's now and immutable object by way of the constructor.  And you stop having to tell people "Hey you are doing it wrong, this is the TRUE way it's ment to be done." when these order of operations Issues come up every now and again.

**Comment by jbogard** [comment]
Well, to be fair the order of operations is still a problem BUT now I have a much easier time fixing it.

**Comment by thomaslevesque** [comment]
> I don't think that's a huge deal, but I could expose a static setter somewhere:
> 
>     var builder = config.CreateExpressionBuilder();
>     AutoMapper.QueryableExtensions.Factories.ExpressionBuilderFactory = () => builder;
>     var dtos = dbContext.Users.ProjectTo<UserDto>().ToList();

I don't think this is a good idea. Using a static for this only means more thread safety issues. Of course you could make it `[ThreadStatic]`, but it would still be unsafe, because any method that you call on the same thread could overwrite it. Passing the builder to `ProjectTo` is much cleaner IMO.

**Comment by TylerCarlson1** [comment]
I think `MapperConfiguration` & `Profile` should hide `IConfiguration` & `IProfileExpression` interfaces explicitly.  This way they should be less likely to change after constructed.

**Comment by TylerCarlson1** [comment]
Ok hiding interfaces by explicitly inheriting from them, looks like it's going to mess up existing working code.  Maybe this is a 5.x thing when all the Obsolete code goes away.  Don't know you how you can obsolete the functions now, other than shadowing the interface functions.

**Comment by lbargaoanu** [comment]
Maybe BuildMapper instead of CreateMapper?
The cleaner way is for the MapperConfiguration to be read-only after the constructor. Seal should not be exposed, should be implicit in the constructor. ProjectTo<Foo>(config) sounds better than exposing the ExpressionBuilder. Why should the user know about a class it doesn't really use, doesn't call its members. A builder would make sense if it did some init work in CreateBuilder, but it seems it doesn't right now.

**Comment by lbargaoanu** [comment]
If existing tests don't run as they are, according to semantic versioning the next version should be 5.0

**Comment by dagostinelli** [comment]
[I just read Jimmy's Blog article about this ] (https://lostechies.com/jimmybogard/2016/01/21/removing-the-static-api-from-automapper) and I **really** love this lib so I'm chiming in with my radio-caller, vaguely related to currently conversation, viewpoint.

The whole "get rid of the static API" thing isn't really a big issue for me.  Jimmy's reasons make sense.  So, "Go for it!".  Make the internals simple and thread-safe.  :+1: 

All that other stuff about how best to use Automapper with Linq, changing configuration settings, expression builder, factories, ORMs, etc -- I'm not so much interested in any of that.  

I more or less restrict my use of AutoMapper to 2 uses cases. 1) to go from IDataReader to a list of DTO's. 2) to map Models to ViewModels and back.  I've written very large apps and smallish apps with those use cases.  I want to advocate for those use cases to be kept/made extremely lean.

I should also mention that I'm a member of that certain class of programmers who rips out dependency injection and the ORMs (and can still unit test just fine).  That likely explains why I never use the Automapper Linq stuff, configurations, or even a complex Map.  So I won't pay attention to much of it in the API and I'm not going to advocate for it.  I realize that others use it.  Please balance this.

So with my 2 simple use cases in mind...

The convention I enjoy with Automapper is quite simply that Automapper is a black box and it just works.  I don't want to dig around with Configurations or Factories or Expression Builders.

In fact, I wish we could use implied maps for those prototyping and throwaway projects (and Automapper could internally create the maps and cache them)

```
var mapper = new Mapper();
var destinationObject = mapper.Map<SourceType, DestinationType>(sourceObject);
```
And then for those production projects where we really should call CreateMap:
``` 
var mapper = new Mapper();
mapper.MustCallCreateMap = true; // I'm asking to invert the current default behavior
var destinationObject = mapper.Map<SourceType, DestinationType>(sourceObject);
// exception thrown
```
That is the sum-total of how I want to interact with AutoMapper.  It covers 95% of my code on all the many projects that I have used AutoMapper for.  

Building off of what I wrote above, when we need to get involved with a factory or configurations for those more complex, advanced use cases, I would like to advocate for these variations:
```
var mapper = new Mapper(); //default configuration
mapper.Configuration.whateverX
mapper.Configuration.whateverY
mapper.Configuration.whateverZ
// or
var mapper = new Mapper(configuration)
```
For those advanced use-cases, I'm really not speaking as a user of an advanced configuration. 

When I saw mention of the words, Factory and Configuration in the snippets about this enhancement, I thought I should throw in my 2-cents and advocate for a cleaner path.  I was looking at [this, in part](https://gist.github.com/jbogard/a60ba6a0230d962c3a7c#file-newlinq-cs) as well as some of the snippets above.

Many thanks for the great library.  I hope to someday meet you in person Jimmy.

**Comment by jbogard** [comment]
@dagostinelli Oh that's interesting. I will need to make the configuration more or less immutable, so today you can do this:

```
var config = new MapperConfiguration(cfg => cfg.CreateMissingTypeMaps = true);
var mapper = config.CreateMapper();
```

But it would not be hard to enable this:

```
var mapper = new Mapper(cfg => cfg.CreateMissingTypeMaps = true);
mapper.Map<Foo, Bar>();
```

You can't create new maps, but internally AutoMapper would (we now do this via the conventions support @TylerCarlson1 added).

But I wouldn't allow arbitrary modification of configuration after the fact, that's where all these threading/order of operations problems comes from.

**Comment by dagostinelli** [comment]
Great! That sounds like a nice balance.

**Comment by TylerCarlson1** [comment]
@dagostinelli The default behavior now is must call create map is false.

So for prototyping you would write code like this
```
var mapper = new Mapper(cfg => cfg.CreateMissingTypeMaps = true);
```
It adds a convention saying map everything to everything.

**Comment by abatishchev** [comment]
1. Go for it! I was thinking about exactly the same matter on Friday and very happy to see the progress
2. Version should be bumped up to 5.0
3. I like the idea to simplify interfaces, especially for mocking and unit testing.
Currently everything what happens before IMappingEngine.Map<>() doesn't matter since down the road you deal with its return value only. For us this call is somewhere in the middle of a pipeline
4. As part of this efforts, please simplify DI container registration. Currently this is s hive piece of boilerplate code
5. Also configuration isn't always threead-safe

**Comment by jbogard** [comment]
@abatishchev 

I'm trying to avoid a major version bump by not including any breaking API changes. I'd like to wait for the 5.0 release to remove the static API altogether, and deprecate it first. This will let me get feedback on the non-static API and make changes to it for 5.0

DI should be simpler now, here's one example:
https://github.com/AutoMapper/AutoMapper/blob/develop/src/AutoMapperSamples/InversionOfControl.cs

I'll post a more complete example.

**Comment by abatishchev** [comment]
@jbogard, right. If you just depricate them it's fine to do not, just see how it goes, and then remove and bump. Sounds like a plan!

Thanks for the link about DI simplification.

**Comment by long2know** [comment]
I have no qualms with the new approach, but some things do not work the same and was wondering if you could provide some guidance.

One aspect that I am unsure as I migrate code from 3.x is mapping custom collections. In particular, I have a PaginatedList<T> which inherits from List<T>, but includes paging properties like page number, sortby, etc.

The old way to deal with this was to have an extension method that would map the custom members and then use AfterMap to map the enumerable elements.

It looked something like this:

```
Mapper.CreateMap<Type1, Type2>().IncludePagedResultMapping();

public static IMappingExpression<TSrc, TDest> IncludePagedResultMapping<TSrc, TDest>(this IMappingExpression<TSrc, TDest> expression)
{
  Mapper.CreateMap<PaginatedList<TSrc>, PaginatedList<TDest>>()
.ForMember(dest => dest.PageNumber, opt => opt.MapFrom(src => src.PageNumber))
.ForMember(dest => dest.SortBy, opt => opt.MapFrom(src => src.SortBy))
.AfterMap((s, d) => Mapper.Map<List<TSrc>, List<TDest>>(s, d));

return expression;
}
```

However, changing this to create a Mapper in the AfterMap (`.AfterMap(s,d) => _config.CreateMapper().Map(....)`) causes a stackoverflow.

Is there a preferred, working approach to these types of scenarios in 4.x?  I looked at the wiki regarding Open Generics, and that didn't seem to be the right path.

**Comment by jbogard** [comment]
@long2know can you open a new issue?

**Comment by zpul** [comment]
Please take a look at this question:
https://stackoverflow.com/questions/37319320/non-static-automapper-and-asp-net-mvc

Thanks.

**Comment by johnhwright** [comment]
Just updated to 5.0.2 and the static Mapper.Map methods are still in place (and not obsolete like they were in v4), were these not removed after all? 

If I've configured multiple instances, what exactly is Mapper.Map using?

**Comment by jbogard** [comment]
No they were not. You can use the static API or the instance API. It's your
choice.

**Comment by pcampbell-software** [comment]
Hi,

Can you clarify when it is expected that the static API will be removed? We're moving some legacy code over to using AutoMapper and some previous work in this area uses the `Mapper.Map` syntax. I'd like to avoid the technical debt of refactoring so wondered what kind of time-frame we have?

**Comment by TylerCarlson1** [comment]
This has changed.  Static API is still there along with instances.  The configuration just has to be done once on initialize.

**Comment by pcampbell-software** [comment]
> This has changed. Static API is still there along with instances. The configuration just has to be done once on initialize.

Thank you -- do you know, is it now not intended to be removed at all?

**Comment by TylerCarlson1** [comment]
It will stay.  The issues with it have been resolved, so there's no reason to get rid of it, as it just wraps around a single instance now.

**Comment by pcampbell-software** [comment]
> It will stay. The issues with it have been resolved, so there's no reason to get rid of it, as it just wraps around a single instance now.

Thank you, this will simplify matters for us.
